packages/material-ui-lab/src/TreeView/TreeView.js

```diff
@@ -207,75 +222,115 @@ const TreeView = React.forwardRef(function TreeView(props, ref) {
+        base = base.filter(id => id === start || id !== previousArrowSelection.current);
+      } else {
+        base.push(value);
       }
+    } else {
+      base.push(value);
     }
 
-    if (flag && event) {
-      event.preventDefault();
-      event.stopPropagation();
+    previousArrowSelection.current = value;
+
+    if (onNodeSelect) {
+      onNodeSelect(event, base);
     }
+
+    setSelectedState(base);
   };
 
-  const getIndexFirstChars = (firstChars, startIndex, char) => {
-    for (let i = startIndex; i < firstChars.length; i += 1) {
-      if (char === firstChars[i]) {
-        return i;
-      }
+  const handleRangeSelect = (event, start, end) => {
+    let base = selected;
+    // If last selection was a range selection ignore nodes that were selected.
+    if (lastSelectionMode.current === 'RANGE') {
+      base = selected.filter(id => lastRangeSelection.current.indexOf(id) === -1);
     }
-    return -1;
-  };
 
-  const setFocusByFirstCharacter = (id, char) => {
-    let start;
-    let index;
-    const lowercaseChar = char.toLowerCase();
+    const range = getNodesInRange(start, end);
+    lastRangeSelection.current = range;
+    let newSelected = base.concat(range);
+    newSelected = newSelected.filter((id, i) => newSelected.indexOf(id) === i);
 
-    const firstCharIds = [];
-    const firstChars = [];
-    // This really only works since the ids are strings
-    Object.entries(firstCharMap.current).forEach(([nodeId, firstChar]) => {
-      const map = nodeMap.current[nodeId];
-      const visible = map.parent ? isExpanded(map.parent) : true;
+    if (onNodeSelect) {
+      onNodeSelect(event, newSelected);
+    }
 
-      if (visible) {
-        firstCharIds.push(nodeId);
-        firstChars.push(firstChar);
-      }
-    });
+    setSelectedState(newSelected);
+  };
 
-    // Get start index for search based on position of currentItem
-    start = firstCharIds.indexOf(id) + 1;
-    if (start === nodeMap.current.length) {
-      start = 0;
+  const handleMultipleSelect = (event, value) => {
+    let newSelected = [];
+    if (selected.indexOf(value) !== -1) {
+      newSelected = selected.filter(id => id !== value);
+    } else {
+      newSelected = [value, ...selected];
     }
 
-    // Check remaining slots in the menu
-    index = getIndexFirstChars(firstChars, start, lowercaseChar);
+    if (onNodeSelect) {
+      onNodeSelect(event, newSelected);
+    }
 
-    // If not found in remaining slots, check from beginning
-    if (index === -1) {
-      index = getIndexFirstChars(firstChars, 0, lowercaseChar);
+    setSelectedState(newSelected);
+  };
+
+  const handleSingleSelect = (event, value) => {
+    const newSelected = multiSelect ? [value] : value;
+    setSelectedState(newSelected);
+
+    if (onNodeSelect) {
+      onNodeSelect(event, newSelected);
     }
 
-    // If match was found...
-    if (index > -1) {
-      focus(firstCharIds[index]);
+    setSelectedState(newSelected);
```

Comment:
It's called before and after onNodeSelect

Reply:
Oh, nice spot!